### DISCUSSION TITLE
question about the linear objective term in the MPC example

### DISCUSSION THREAD

**Discussion by Jushan-Randy-Chen** [osqp/osqp#540] [Mathematical Modeling] [answered]
I found the MPC example on the documentation page very helpful. However, I am unsure about the presence of the linear objective term here:

```
......
# - linear objective
q = np.hstack([np.kron(np.ones(N), -Q@xr), -QN@xr, np.zeros(N*nu)])
......
```

In the problem description on the very same page, the MPC problem is given as:
![image](https://github.com/osqp/osqp/assets/56526142/5d38f241-f1b5-44e4-9490-4c19a1ce1283)

Therefore, why is there a linear objective `q` instead of a quadratic terminal cost?

**Comment by goulart-paul** [comment, marked as answer]
I think you are referring to the example [here](https://osqp.org/docs/examples/mpc.html).

The objective function in that example (both linear and quadratic parts) is 
```python 
# - quadratic objective
P = sparse.block_diag([sparse.kron(sparse.eye(N), Q), QN,
                       sparse.kron(sparse.eye(N), R)], format='csc')
# - linear objective
q = np.hstack([np.kron(np.ones(N), -Q@xr), -QN@xr, np.zeros(N*nu)])
```
The objective function is minimizing the (weighted) square of the difference between $x$ and some reference state $x_r$, the latter of which is not a decision variable.   Terms like $\frac{1}{2}(x-x_r)^TQ(x-x_r)$ therefore have both a quadratic component $\frac{1}{2}x^TQx$ and a linear part $\frac{1}{2}(x_r^TQx + x^TQx_r) = (Qx_r)^Tx$.   Note that the last step works because $Q$ is symmetric.

There will also be constant part $\frac{1}{2}x_r^TQx_r$, but we drop that because it doesn't change the optimizer.

**Reply by Jushan-Randy-Chen** [reply]
Thank you so much!
I have a follow-up question: I an doing a multi-agent MPC problem in the context of trajectory planning, and thus I want to add a collision avoidance cost to the objective function (either quadratic in the distance between any two agents `i,j` or a logarithmic barrier function). Is there any clever way to acheive that by simply transforming the matrix `P`?

**Comment by goulart-paul** [comment]
You can't include an objective term that maximizes the distance between vectors since that would be non-convex.  The solver will throw an error saying that your objective function has a negative eigenvalue.   The solver will also not support a log barrier function -- only (convex) quadratic objectives are supported.

If you really want either of those things then you need a solver like IPOPT that supports non-convex problems.

**Reply by Jushan-Randy-Chen** [reply]
Thank you! I guess I'll try linearizing the 2-norm distance to get a relaxed constraint
